fe/src/main/java/org/apache/doris/analysis/GroupByClause.java

```diff
@@ -0,0 +1,321 @@
+                    }
+                }
+                break;
+            case GROUPING_SETS:
+                if (groupingSetList != null) {
+                    strBuilder.append(" GROUPING SETS (");
+                    boolean first = true;
+                    for(List<Expr> groupingExprs : groupingSetList) {
+                        if (first) {
+                            strBuilder.append("(");
+                            first = false;
+                        } else {
+                            strBuilder.append(", (");
+                        }
+                        for (int i = 0; i < groupingExprs.size(); ++i) {
+                            strBuilder.append(groupingExprs.get(i).toSql());
+                            strBuilder.append((i + 1 != groupingExprs.size()) ? ", " : "");
+                        }
+                        strBuilder.append(")");
+                    }
+                    strBuilder.append(")");
+                }
+                break;
+            case CUBE:
+                if (groupingExprs != null) {
+                    strBuilder.append("CUBE (");
+                    for (int i = 0; i < groupingExprs.size(); ++i) {
+                        strBuilder.append(groupingExprs.get(i).toSql());
+                        strBuilder.append((i + 1 != groupingExprs.size()) ? ", " : "");
+                    }
+                    strBuilder.append(")");
+                }
+                break;
+            case ROLLUP:
+                if (groupingExprs != null) {
+                    strBuilder.append("ROLLUP (");
+                    for (int i = 0; i < groupingExprs.size(); ++i) {
+                        strBuilder.append(groupingExprs.get(i).toSql());
+                        strBuilder.append((i + 1 != groupingExprs.size()) ? ", " : "");
+                    }
+                    strBuilder.append(")");
+                }
+                break;
+            default:
+                break;
+        }
+        return strBuilder.toString();
+    }
+
+    @Override
+    public GroupByClause clone() {
+        return new GroupByClause(this);
+    }
+
+    public void reset() {
+        if (groupingExprs != null) Expr.resetList(groupingExprs);
+        this.analyzed_ = false;
+    }
+
+    public boolean isEmpty() {
+        return groupingExprs == null || groupingExprs.isEmpty();
+    }
+
+    private void addGroupingId(Analyzer analyzer) throws  AnalysisException {
+        TupleDescriptor tupleDesc = analyzer.getDescTbl().createTupleDescriptor(GROUPING__ID);
+        groupingIdSlotRef = new VirtualSlotRef(GROUPING__ID, Type.BIGINT, tupleDesc);
+        groupingIdSlotRef.analyze(analyzer);
+        groupingExprs.add(groupingIdSlotRef);
+    }
+
+    private void buildGroupingClause(Analyzer analyzer) throws AnalysisException {
+        groupingIdList = new ArrayList<>();
+        BitSet bitSetAll = new BitSet();
```

Comment:
I think it is better to init groupingIdList in construct function,  it can avoid to get null by others and conforms groupingIdList is inited once.